### ISSUE TITLE
diag(p3): TransformerPolicyNetwork BC-fit (Phase 1.5 escalation if #278+#279 fail)

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/bucket-brigade#281]
## Hypothesis being tested

If both class-balanced BC (#278) and hidden_size sweep (#279) fail to recover the WORK-house discriminating sub-task, the architectural gap isn't pure capacity — it's *kind*: the MLP may not have the inductive biases needed to route `agent_identity → owned-house-index → houses[that_index] is burning?`. A Transformer's attention mechanism handles this kind of indexed-lookup naturally.

## Context

A `TransformerPolicyNetwork` already exists in `bucket_brigade/training/networks.py` (per PR #276 builder's note). This is a 1-line architecture swap.

## Protocol

Run BC-fit (class-balanced if #278 indicated it helps) with TransformerPolicyNetwork instead of MLP. Compare against the best MLP from #279.

## Verdict logic

| TransformerPolicyNetwork on house-on-WORK | Interpretation |
|---|---|
| **≥ 0.90** | Attention solved it. PPO's MLP was the wrong inductive bias. Next: full PPO with Transformer. |
| **No better than best MLP** | Architecture wasn't the gap. Specialist may use information that's not in the obs — fundamental task reformulation needed. |
| **Modest improvement** | Combined approach (Transformer + bigger hidden_size + class balancing) might work. File production training. |

## Cost

~1 hour. Same protocol as #272/#278/#279 with different network class.

## Dependency

**Blocked on #279.** Don't run if MLP+capacity already solves it.

## Acceptance criteria

- [ ] BC with Transformer policy net
- [ ] Compared against #279 best MLP
- [ ] Verdict applied; production-PPO follow-up filed if successful

## References

- PR #276 (#272 verdict)
- #278, #279 (Phase 1.5 sequence — must complete first)
- `bucket_brigade/training/networks.py::TransformerPolicyNetwork` (existing, ready to use)

---

### Status update: this issue is likely deprecated by #278's cross-finding

PR #278 (merged 2026-05-17) demonstrated that the **default MLP `PolicyNetwork(hidden_size=64)`** reaches BC `gap_closed = 0.934` at 40k pairs/agent × 30 epochs. This directly contradicts #272's `INDUCTIVE_BIAS_GAP` verdict and implies the MLP *does* have the inductive bias to route `agent_identity → owned-house-index → houses[that_index]`. The Transformer-escalation premise (capacity/architecture is the bottleneck) is no longer the leading hypothesis.

**Recommended scope** (in priority order):

1. **Defer until #279 resolves.** If class-balanced BC (#279) confirms #278's 0.934 result holds with the original (10k/10ep) budget after rebalancing, this issue should be closed as resolved (mark `loom:blocked` flag below stays until human decides).
2. **If #279 *and* the (now-likely-deprecated) #280 hidden_size sweep both come back negative on house-on-WORK accuracy**, then escalate to Transformer per the original protocol below.
3. **Orthogonal motivation**: even if not needed diagnostically, Transformer vs MLP on full PPO production is a legitimate comparison — but that's a separate issue, not this one.

### Dependencies

- [x] #278: BC-init + PPO continuation (MERGED — cross-finding documented)
- [ ] #279: Class-balanced BC-fit (open, curating) — **must complete before this issue runs**
- [ ] #280: hidden_size sweep (open, curating) — likely deprecated by #278's finding but listed for completeness

This issue cannot proceed until #279 returns a verdict. If #279 closes the capacity-vs-imbalance question affirmatively (MLP solves house-on-WORK with rebalancing), this issue should be closed without execution.

### Problem statement

If both class-balanced BC (#279) and hidden_size capacity (#280) fail to recover the WORK-house discriminating sub-task at adequate budget, the architectural gap is *kind*, not *size*: an MLP may lack the inductive biases to route `agent_identity → owned-house-index → houses[that_index] is burning?`. A Transformer's self-attention handles indexed lookup naturally. This is a Phase 1.5 escalation, gated on #279/#280 failure.

### Hypothesis under test

`BC(TransformerPolicyNetwork)` reaches house-on-WORK accuracy ≥ 0.90 on the same demonstrations that the MLP fails to fit, isolating attention-based indexed lookup as the missing inductive bias.

### Implementation guidance

`TransformerPolicyNetwork` already exists at `bucket_brigade/training/networks.py:324` with this signature:

```python
TransformerPolicyNetwork(
    obs_dim: int,
    action_dims: List[int],
    num_houses: int = None,        # auto-detected from obs_dim
    house_features: int = 4,        # fire_level, bucket_count, distance, threatened
    d_model: int = 256,             # transformer embedding dim
    nhead: int = 4,                 # attention heads
    num_layers: int = 3,            # transformer encoder layers
    dim_feedforward: int = 512,
    dropout: float = 0.1,
)
```

~350K params at defaults vs ~14K for `PolicyNetwork(hidden_size=64)`.

**Architectural fit (verify before running)**: confirm the obs layout fed to `bc_fit_only.py` (post-#204 identity-tail one-hot) matches the Transformer's assumed `[houses × 4 features + global]` layout. The Transformer auto-detects `num_houses` from `obs_dim`; if global features (identity tail, etc.) make `obs_dim % house_features != 0` the remainder is routed through `global_embed`. Sanity-check the parse: with `num_agents=2`, post-#204 obs ≈ `num_houses*4 + num_agents` so a 10-house env has `obs_dim = 42`, `num_houses = 10`, `global_dim = 2`. Good.

**Protocol** (mirror #279's, swap network class only):

1. Reuse `bc_fit_only.py`'s data path and verdict harness; add a `--network {mlp,transformer}` CLI flag (or write a `bc_fit_transformer.py` sibling — builder discretion).
2. Use whichever BC budget produced #279's best MLP result (likely 40k pairs / 30 epochs at minimum).
3. If #279 found class balancing helps, apply the same weighting here.
4. Compare house-on-WORK accuracy and `gap_closed` against #279's best MLP cell.
5. Same 3-seed protocol as the diagnostic ladder.

### Verdict logic

| Transformer house-on-WORK | Interpretation | Next step |
|---|---|---|
| ≥ 0.90 *and* clearly beats MLP | Attention solved indexed lookup. PPO's MLP was wrong inductive bias. | File PPO-with-Transformer production issue. |
| Within ~0.05 of best MLP | Architecture wasn't the gap. Specialist may use info not in obs — task reformulation needed. | File obs/reward redesign issue (env-side intervention per project thesis). |
| Modest improvement (0.05–0.15) | Combined approach (Transformer + balancing + budget) plausible but inconclusive. | Decide based on cost; default = file env-side issue. |

### Affected Files

- `experiments/p3_specialization/bc_fit_only.py:164` — current MLP instantiation; add `--network` flag here or fork into a transformer-specific module.
- `experiments/p3_specialization/bc_init.py:128` — sibling BC harness; same instantiation pattern.
- `bucket_brigade/training/networks.py:324` — `TransformerPolicyNetwork` (use as-is; no edits expected).
- `bucket_brigade/training/__init__.py:8` — already re-exports `TransformerPolicyNetwork`.
- `experiments/marl/train_gpu.py:280` — reference for an existing Transformer instantiation pattern (don't import from here; just match the constructor call).
- `experiments/p3_specialization/diagnostics/results/` — write results JSON alongside #278/#279 artifacts.

### Test Plan

- [ ] Sanity: parameter count for `TransformerPolicyNetwork(obs_dim=42, action_dims=[10,2,2])` matches docstring's ~350K (within 2×).
- [ ] Forward-pass shape check: instantiate net, run one batch through, assert action-head logits have shape `(B, 10)`, `(B, 2)`, `(B, 2)`.
- [ ] Reproducibility: 3 seeds; report mean ± stdev for `gap_closed` and per-head accuracies.
- [ ] Verdict cell applied per table above; results JSON committed to `experiments/p3_specialization/diagnostics/results/`.
- [ ] Cross-comparison: side-by-side house-on-WORK accuracy vs #279's best MLP cell.
- [ ] Cost gate: experiment runs in < 2 hours on `COMPUTE_HOST_PRIMARY`; abort and re-scope if blown.

### References

- PR #278 (cross-finding: MLP @ 64-hidden, 40k/30ep → `gap_closed = 0.934`)
- PR #276 / #272 (original `INDUCTIVE_BIAS_GAP` verdict, now contradicted)
- #279 (class-balanced BC; primary blocker)
- #280 (hidden_size sweep; secondary blocker)
- `bucket_brigade/training/networks.py:324` (`TransformerPolicyNetwork`)
- `bucket_brigade/training/networks.py:13` (`PolicyNetwork` baseline)

**Comment by rjwalters** [comment]
## Cross-issue finding from PR #278: Transformer BC may be unnecessary

PR #278's builder demonstrated that the default MLP (`PolicyNetwork(hidden_size=64)`) reaches BC `gap_closed = 0.934` with an adequate training budget (40k pairs, 30 epochs). The MLP **does** have the inductive bias to route `agent_identity → owned-house-index → houses[that_index]`; #272's failure was a budget issue, not a representational one.

## Recommendation for whoever builds this

Consider closing this issue as "resolved by PR #278 cross-finding" unless:
- #279 (class-balanced BC) and #280 (hidden_size sweep) BOTH come back negative, indicating MLP fundamentally can't reach high WORK-house accuracy regardless of budget or balancing
- You specifically want to compare Transformer vs MLP for PPO production runs (orthogonal motivation, not a diagnostic)

The original motivation — "MLP can't represent the specialist" — is now contradicted by direct evidence.

## References
- PR #278 (the cross-finding source)
- PR #276 (#272 verdict)
- Judge of #276

**Comment by rjwalters** [comment]
Curating this issue per user request

**Comment by rjwalters** [comment]
Curation complete.

**Key updates**:
- Enhanced body with explicit dependency framing: blocked on #279 (and #280, though #280 is likely deprecated by #278's cross-finding).
- Added prominent status note that PR #278 found MLP @ default `hidden_size=64`, 40k/30ep reaches `gap_closed = 0.934`, contradicting the original `INDUCTIVE_BIAS_GAP` premise. Recommended this issue close-without-execution if #279 confirms.
- Cited `TransformerPolicyNetwork` at `bucket_brigade/training/networks.py:324` with full config (d_model=256, nhead=4, num_layers=3, dim_feedforward=512, ~350K params).
- Verified obs-layout fit (post-#204 identity tail → `obs_dim=42, num_houses=10, global_dim=2`).
- Added Affected Files, Test Plan, refined verdict table with concrete next-step issue routing per cell.
- Added cost gate (< 2 hours on `COMPUTE_HOST_PRIMARY`).

**Labels**: `loom:curated` + `loom:blocked` (gated on #279 verdict; do NOT mark `loom:issue` until #279 closes and the human confirms escalation is still warranted).

**Duplicate check**: `check-duplicate.sh` flagged #279 at 100% (shared "Phase 1.5" terminology only) — scope is distinct (Transformer escalation vs. class-balanced MLP). Not a duplicate.
